numpyro/contrib/module.py

```diff
@@ -429,3 +431,381 @@ def random_haiku_module(
+
+        # Create our safe Rngs object
+        rngs = SafeRngs(**rngs_dict)
+
+        # Handle initialization based on input_shape or args
+        if input_shape is not None:
+            # Create a dummy input for initialization
+            dummy_input = jnp.ones(input_shape)
+
+            # Create module instance
+            model = nn_module(*args, rngs=rngs, **kwargs)
+
+            # Call the model to initialize parameters
+            if apply_rng and "dropout" in apply_rng:
+                dropout_rngs = {"dropout": rngs_dict["dropout"]}
+                model(dummy_input, rngs=dropout_rngs)
+            else:
+                model(dummy_input)
+        else:
+            # Initialize without dummy input
+            model = nn_module(*args, rngs=rngs, **kwargs)
+
+        # Extract parameters
+        param_dict = {}
+        for name_attr, var in model.__dict__.items():
+            if isinstance(var, nnx.Param):
+                param_dict[name_attr] = var.value
+            elif hasattr(var, "__dict__"):
+                # Handle nested modules like BatchNorm
+                for sub_name, sub_var in var.__dict__.items():
+                    if isinstance(sub_var, nnx.Param):
+                        param_dict[f"{name_attr}.{sub_name}"] = sub_var.value
+
+        # Register parameters with NumPyro
+        if param_dict:
+            numpyro.param(module_key, param_dict)
+            module_params = param_dict
+
+        # Handle mutable state if needed
+        if mutable:
+            module_state = {m: {} for m in mutable}
+
+            # Extract mutable state
+            for name_attr, var in model.__dict__.items():
+                if hasattr(var, "__dict__"):
+                    for sub_name, sub_var in var.__dict__.items():
+                        for m in mutable:
+                            if m in sub_name and not isinstance(sub_var, nnx.Param):
+                                module_state[m][f"{name_attr}.{sub_name}"] = sub_var
+
+            if any(module_state.values()):
+                numpyro_mutable(name + "$state", module_state)
+
+    # Define the apply function using JAX's functional approach
+    def apply_fn(*call_args, **call_kwargs):
+        # Get user-provided RNGs or create new ones
+        user_rngs = call_kwargs.pop("rngs", None)
+        rng_key = numpyro.prng_key()
+
+        # Create Rngs object for this call
+        if user_rngs is not None:
+            if isinstance(user_rngs, dict):
+                rngs = SafeRngs(**user_rngs)
+            elif hasattr(user_rngs, "_fields"):
+                # Convert to our safe version
+                rngs_dict = {key: getattr(user_rngs, key) for key in user_rngs._fields}
+                rngs = SafeRngs(**rngs_dict)
+            else:
+                rngs = SafeRngs(params=rng_key)
+        else:
+            rngs = SafeRngs(params=rng_key)
+
+        # Create a new module instance
+        model = nn_module(*args, rngs=rngs, **kwargs)
```

Comment:
I think for `nnx`, the module is initialized before providing to numpyro.module. Hence, to update params, you can do something like
```
graph_def, state = nnx.split(nn_module)
nnx.replace_by_pure_dict(state, process_raw_dict(module_params))
nn_module = nnx.merge(graph_def, state)
```
like in https://flax.readthedocs.io/en/latest/guides/surgery.html